client/tracing/src/block/mod.rs

```diff
@@ -0,0 +1,314 @@
+		let spans = Mutex::new(HashMap::new());
+		let events = Mutex::new(Vec::new());
+		let block_subscriber = BlockSubscriber::new(targets, spans, events);
+		let dispatch = Dispatch::new(block_subscriber);
+
+		{
+			let dispatcher_span = tracing::debug_span!(target: "state_tracing",
+				"execute_block", extrinsics_len = block.extrinsics().len());
+			let _guard = dispatcher_span.enter();
+			if let Err(e) = dispatcher::with_default(&dispatch, || {
+				let span = tracing::info_span!(
+					target: TRACE_TARGET,
+					"trace_block",
+				);
+				let _enter = span.enter();
+				self.client.runtime_api().execute_block(&parent_id, block)
+			}) {
+				return Err(format!("Error executing block: {:?}", e));
+			}
+		}
+
+		let block_subscriber = dispatch.downcast_ref::<BlockSubscriber>()
+			.ok_or("Cannot downcast Dispatch to BlockSubscriber after tracing block")?;
+		let mut span_data: Vec<SpanDatum> = block_subscriber.spans
+			.lock()
+			.drain()
+			.map(|(_, s)| SpanDatum::from(s))
+			.collect::<Vec<SpanDatum>>();
+
+		// Is there a way to do this sort without collecting twice? My only idea was to add start_time
+		// into `Span` - zeke
+		span_data.sort_by(|a, b| a.start_time.cmp(&b.start_time));
+
+		let spans: Vec<Span> = span_data
+			.into_iter()
+			// Patch wasm identifiers
+			.filter_map(|s| patch_and_filter(s, targets))
+			.collect();
+
+		let events: Vec<_> = block_subscriber.events
+			.lock()
+			.drain(..)
+			.filter(|e| event_key_filter(e, storage_keys))
+			.map(|s| s.into())
+			.collect();
+
+		tracing::debug!(target: "state_tracing", "Captured {} spans and {} events", spans.len(), events.len());
+
+		let block_trace = BlockTrace {
+			block_hash: block_id_as_string(id),
+			parent_hash: block_id_as_string(parent_id),
+			tracing_targets: targets.to_string(),
+			storage_keys: storage_keys.to_string(),
+			spans,
+			events,
+		};
+
+		let block_trace_size = serde_json::to_vec(&block_trace)
+			.map_err(|_| "Failed to serialize payload".to_string())?
+			.len();
+		let response = if block_trace_size  > MAX_PAYLOAD {
+			TraceBlockResponse::TraceError(TraceError {
+				error:
+					format!(
+						"Payload was {} bytes; it must be less than {} bytes",
+						block_trace_size,
+						MAX_PAYLOAD
+					)
+			})
+		} else {
+			TraceBlockResponse::BlockTrace(block_trace)
+		};
+
+		Ok(response)
+	}
+}
+
+fn event_key_filter(event: &TraceEvent, storage_keys: &str) -> bool {
+	event.values.string_values.get("key")
+		.and_then(|key| Some(check_target(storage_keys, key, &event.level)))
+		.unwrap_or(false)
+}
+
+fn patch_and_filter(mut span: SpanDatum, targets: &str) -> Option<Span> {
+	if span.name == WASM_TRACE_IDENTIFIER {
+		span.values.bool_values.insert("wasm".to_owned(), true);
+		if let Some(n) = span.values.string_values.remove(WASM_NAME_KEY) {
+			span.name = n;
+		}
+		if let Some(t) = span.values.string_values.remove(WASM_TARGET_KEY) {
+			span.target = t;
+		}
+		if !check_target(targets, &span.target, &span.level) {
```

Comment:
Can we do this check first so we can exit early?

Reply:
I don't think that would be any benefit because we still need to get the name and target from the values, which is all that's happening above really

Reply:
Lines 313-315 could be moved below this check.